apps/rush-lib/src/logic/buildCache/ProjectBuildCache.ts

```diff
@@ -0,0 +1,220 @@
+              if (!projectsThatHaveBeenProcessed.has(dependency)) {
+                newProjectsToProcess.add(dependency);
+              }
+            }
+          }
+        }
+
+        projectsToProcess = newProjectsToProcess;
+      }
+
+      const sortedProjectStates: string[] = projectStates.sort();
+      const hash: crypto.Hash = crypto.createHash('sha1');
+      const serializedOutputFolders: string = JSON.stringify(this._projectOutputFolderNames);
+      hash.update(serializedOutputFolders);
+      hash.update(RushConstants.hashDelimiter);
+      hash.update(this._command);
+      hash.update(RushConstants.hashDelimiter);
+      for (const projectHash of sortedProjectStates) {
+        hash.update(projectHash);
+        hash.update(RushConstants.hashDelimiter);
+      }
+
+      this.__cacheId = hash.digest('hex');
+    }
+
+    return this.__cacheId;
+  }
+
+  public constructor(options: IProjectBuildCacheOptions) {
+    this._project = options.projectBuildCacheConfiguration.project;
+    this._command = options.command;
+    this._buildCacheProvider = options.buildCacheProvider;
+    this._packageChangeAnalyzer = options.packageChangeAnalyzer;
+    this._projectOutputFolderNames = options.projectBuildCacheConfiguration.projectOutputFolders;
+    this._terminal = options.terminal;
+  }
+
+  public async tryHydrateFromCacheAsync(): Promise<boolean> {
+    const cacheId: string | undefined = this._cacheId;
+    if (!cacheId) {
+      this._terminal.writeWarningLine('Unable to get cache ID. Ensure Git is installed.');
+      return false;
+    }
+
+    const cacheEntryBuffer:
+      | Buffer
+      | undefined = await this._buildCacheProvider.tryGetCacheEntryBufferByIdAsync(this._terminal, cacheId);
+    if (!cacheEntryBuffer) {
+      this._terminal.writeVerboseLine('No build cache hit.');
+      return false;
+    }
+
+    this._terminal.writeLine('Build cache hit.');
+
+    const projectFolderPath: string = this._project.projectFolder;
+
+    // Purge output folders
+    this._terminal.writeVerboseLine(`Clearing cached folders: ${this._projectOutputFolderNames.join(', ')}`);
+    await Promise.all(
+      this._projectOutputFolderNames.map((outputFolderName: string) =>
+        FileSystem.deleteFolderAsync(path.join(projectFolderPath, outputFolderName))
+      )
+    );
+
+    const tarStream: stream.Writable = tar.extract({ cwd: projectFolderPath });
+    const success: boolean = await new Promise(
+      (resolve: (result: boolean) => void, reject: (error: Error) => void) => {
+        try {
+          tarStream.on('error', (error: Error) => reject(error));
+          tarStream.on('close', () => resolve(true));
+          tarStream.on('drain', () => resolve(true));
+          tarStream.write(cacheEntryBuffer);
+        } catch (e) {
+          reject(e);
+        }
+      }
+    );
+
+    if (success) {
+      this._terminal.writeLine('Successfully hydrated build output from cache.');
```

Comment:
"Hydrate" is kind of a weird term. I know many utilities use that term, but a polished product will typically prefer friendly/intuitive terminology. 

Maybe instead of "hydrate" we could say "restored from cache" or "extracted from cache"?